### ISSUE TITLE
Coverage doesn't work with long classpaths

### ISSUE THREAD

**Issue by viveksjain** [bazelbuild/bazel#11847]
### Description of the problem / feature request:

If the classpath for a test target exceeds the limit (specified at https://github.com/bazelbuild/bazel/blob/fd5d29aa4ef907a15ec6db0e359e86df1acf64fc/src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt#L358) the classpath is written to separate file, and coverage output is empty.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Clone bazel repo and run
```
$ bazel coverage //examples/java-native/src/test/java/com/example/myproject:custom
...
//examples/java-native/src/test/java/com/example/myproject:custom        PASSED in 0.7s
  /home/ubuntu/.cache/bazel/_bazel_ubuntu/5bb6c1fa5da68b7db1883687715859ce/execroot/io_bazel/bazel-out/k8-fastbuild/testlogs/examples/java-native/src/test/java/com/example/myproject/custom/coverage.dat

$ bazel coverage --action_env=CLASSPATH_LIMIT=100 //examples/java-native/src/test/java/com/example/myproject:custom
...
//examples/java-native/src/test/java/com/example/myproject:custom        PASSED in 0.6s
# No .dat file mentioned
$ ls -lh /home/ubuntu/.cache/bazel/_bazel_ubuntu/5bb6c1fa5da68b7db1883687715859ce/execroot/io_bazel/bazel-out/k8-fastbuild/testlogs/examples/java-native/src/test/java/com/example/myproject/custom/coverage.dat
-r-xr-xr-x 1 ubuntu ubuntu 0 Jul 27 16:27 /home/ubuntu/.cache/bazel/_bazel_ubuntu/5bb6c1fa5da68b7db1883687715859ce/execroot/io_bazel/bazel-out/k8-fastbuild/testlogs/examples/java-native/src/test/java/com/example/myproject/custom/coverage.dat
```

### What operating system are you running Bazel on?

Linux

### What's the output of `bazel info release`?

release 3.4.1

###  Have you found anything relevant by searching the web?

No :(

### Any other information, logs, or outputs that you want to share?

Workaround/Pointer to fix would be appreciated. I briefly looked at JacocoCoverageRunner.java but don't have enough context with this code to make progress quickly.

**Comment by jin** [comment]
fyi @comius

**Comment by TheKaur** [comment]
Any update on this? @jin @comius

**Comment by jin** [comment]
I'll defer to @comius for long term plans, but there is nobody assigned to work on this right now.

**Comment by comius** [comment]
@c-mita or @oquenchil might be more familiar with coverage

**Comment by TheKaur** [comment]
@c-mita @oquenchil any plans to fix this soon?

**Comment by TheKaur** [comment]
@c-mita Charles, the fix didn't work. If you try the original repro steps listed in the ticket, you'd observe that the coverage.dat is not created. 

If we supply a longer classpath for our use case, we run into `Argument list too long` error and the test fails.

**Comment by c-mita** [comment]
That sounds like a command line error, that you're hitting `ARG_MAX`. Can you provide the full error?

**Comment by TheKaur** [comment]
Clarification:
If we supply a longer ~~classpath~~ CLASSPATH_LIMIT for our use case, we run into `Argument list too long` error and the test fails.

The error comes from the Tests_wrapper.sh bazel script:
`/home/ubuntu/.cache/bazel/_bazel_ubuntu/718015170c1e1789cf09fd3727c0f670/sandbox/processwrapper-sandbox/6036/execroot/com_rubrik_sdmain/bazel-out/k8-fastbuild/bin/src/java/example/Tests.runfiles/com_rubrik_sdmain/src/java/example/Tests_wrapper.sh: Argument list too long
`

To summarize:
1. Supplying no CLASSPATH_LIMIT results in no coverage.dat.
2. A low CLASSPATH_LIMIT results in no coverage.dat
3. A higher CLASSPATH_LIMIT results in the above error.

**Comment by TheKaur** [comment]
cc @c-mita

**Comment by c-mita** [comment]
Running the example given at the top of this issue does not reproduce any failures:

In the bazel repo:
```
$ bazel coverage --action_env=CLASSPATH_LIMIT=10 //examples/java-native/src/test/java/com/example/myproject:custom --combined_report=lcov
...

$ wc bazel-out/k8-fastbuild/testlogs/examples/java-native/src/test/java/com/example/myproject/custom/coverage.dat
 39  49 990 bazel-out/k8-fastbuild/testlogs/examples/java-native/src/test/java/com/example/myproject/custom/coverage.dat

$ wc bazel-out/_coverage/_coverage_report.dat
 39  49 990 bazel-out/_coverage/_coverage_report.dat
```
Using a custom version of blaze with 94eeea6 reverted results in an empty `coverage.dat` as expected.
If this particular repro is still failing for you then can you provide as many details as you can regarding your setup?


As for your error, it looks like you're using a custom project? I don't know what `Tests_wrapper.sh` is or what it's doing.

Looking at `java_stub_template.txt`, it's seems likely that a sufficiently large classpath and `CLASSPATH_LIMIT` could result in a breach of `ARG_MAX` (as the classpath is just passed to java binary on the command line).

It may also be that a sufficiently large classpath with a low `CLASSPATH_LIMIT` would have also cause problems but this isn't obvious to me and would take me some time to investigate.
